### ISSUE TITLE
CTC lattice failed when using large vocab size

### ISSUE THREAD

**Issue by yuekaizhang** [k2-fsa/k2#982]
[ctc_decode_min.zip](https://github.com/k2-fsa/k2/files/8917328/ctc_decode_min.zip)

The zip file includes a single python file to reproduce. It works when I set len_dict = 500, however, when I increase it to number like 700, it fails. (I think its not gpu OOM issue, since the gpu memory usuage never more than 10Gb, while I have 40gb.)

The log is below:
![image](https://user-images.githubusercontent.com/28823245/174033668-ab9fd59d-d50d-4ab9-a3ef-0a17832d5147.png)

Basically, what I am going to do is using ctc-decoding method in https://github.com/yuekaizhang/icefall/blob/master/egs/librispeech/ASR/conformer_ctc/decode.py

The code looks like below:

![image](https://user-images.githubusercontent.com/28823245/174034390-5a9cb419-8cb9-4bb3-8b1b-110468e704c0.png)

**Comment by pkufool** [comment]
Note, the max_token in `k2.ctc_topo` is the max id in token.txt (i.e. vocab-size - 1), 0 is blank id.

**Comment by yuekaizhang** [comment]
> So you should build the topology with
> 
> ```
> H = k2.ctc_topo(
>   max_token=len_dict - 1,
>   modified=False,
>   device=device,
> )
> ```
> 
> This runs normally in my local machine.

Thanks, would you mind sharing your k2 version? For me, I use `pip3 install k2==1.10.dev20211102+cuda11.0.torch1.7.1 -f https://k2-fsa.org/nightly/`. I modified max_token to len_dict -1 and it reports cuda out of memory this time. (Still using the file in ctc_decode_min.zip)

![image](https://user-images.githubusercontent.com/28823245/174053225-fc245077-9538-456c-931b-abad11053a29.png)

**Comment by pkufool** [comment]
I also got OOM when using GPU, I tested it in CPU.  The log_probs are all zeros, this make it no way to prune out states and arcs, I think it is expected to occur OOM.

**Comment by pkufool** [comment]
1.10 is a very old version, I think.
My version is

```
Collecting environment information...

k2 version: 1.15.1
Build type: Release
Git SHA1: f8d2dba06c000ffee36aab5b66f24e7c9809f116
Git date: Thu Apr 21 12:20:34 2022
Cuda used to build k2: 10.2
cuDNN used to build k2: 8.3.2
Python version used to build k2: 3.8
OS used to build k2: Ubuntu 18.04.5 LTS
CMake version: 3.10.2
GCC version: 7.5.0
CMAKE_CUDA_FLAGS:  --expt-extended-lambda -gencode arch=compute_70,code=sm_70 -D_GLIBCXX_USE_CXX11_ABI=0 --compiler-options -Wall  --compiler-options -Wno-strict-overflow  --compiler-options -Wno-unknown-pragmas
CMAKE_CXX_FLAGS:  -D_GLIBCXX_USE_CXX11_ABI=0 -Wno-unused-variable  -Wno-strict-overflow
PyTorch version used to build k2: 1.10.0+cu102
PyTorch is using Cuda: 10.2
NVTX enabled: True
With CUDA: True
Disable debug: True
Sync kernels : False
Disable checks: False
Max cpu memory allocate: 214748364800
```

**Comment by yuekaizhang** [comment]
> I also got OOM when using GPU, I tested it in CPU. The log_probs are all zeros, this make it no way to prune out states and arcs, I think it is expected to occur OOM.

I see. Thank you very much, I would try to test it with real data.

**Comment by yuekaizhang** [comment]
> Also, standard ctc topology is quadratic with vocab-size, if you have a very big vocabulary, you can try modified ctc topolopy, it won't affect too much to the WERs. Or a better way is to decoded with a HLG graph.

Cool, I was wondering if there is a problem if I use model trained by warp-ctc combining with modified ctc topo to decode. You said it won't affect too much to WER assume we also train the model with modified ctc topo?

**Comment by pkufool** [comment]
> Cool, I was wondering if there is a problem if I use model trained by warp-ctc combining with modified ctc topo to decode. You said it won't affect too much to WER assume we also train the model with modified ctc topo?

According to our previous experiments, you can train with standard ctc and decode with modified ctc topo, only very little degradation on performance.
